packages/protocol/contracts/team/TimeLockTokenPool.sol

```diff
@@ -0,0 +1,269 @@
+    function withdraw(address to, bytes memory sig) external {
+        if (to == address(0)) revert INVALID_PARAM();
+        bytes32 hash = keccak256(
+            abi.encodePacked("Withdraw unlocked Taiko token to: ", to)
+        );
+        address recipient = ECDSAUpgradeable.recover(hash, sig);
+        _withdraw(recipient, to);
+    }
+
+    function getMyGrantSummary(address recipient)
+        public
+        view
+        returns (
+            uint128 amountOwned,
+            uint128 amountUnlocked,
+            uint128 amountWithdrawn,
+            uint128 amountWithdrawable
+        )
+    {
+        Recipient storage r = recipients[recipient];
+        for (uint128 i; i < r.grants.length; ++i) {
+            amountOwned += _getAmountOwned(r.grants[i]);
+            amountUnlocked += _getAmountUnlocked(r.grants[i]);
+        }
+
+        amountWithdrawn = r.amountWithdrawn;
+        amountWithdrawable = amountUnlocked - amountWithdrawn;
+    }
+
+    function getMyGrants(address recipient)
+        public
+        view
+        returns (Grant[] memory)
+    {
+        return recipients[recipient].grants;
+    }
+
+    function _withdraw(address recipient, address to) private {
+        Recipient storage r = recipients[recipient];
+        uint128 amount;
+
+        for (uint128 i; i < r.grants.length; ++i) {
+            amount += _getAmountUnlocked(r.grants[i]);
+        }
+
+        amount -= r.amountWithdrawn;
+        if (amount == 0) revert NOTHING_TO_WITHDRAW();
+
+        r.amountWithdrawn += amount;
+        totalAmountWithdrawn += amount;
+        ERC20Upgradeable(taikoToken).transferFrom(sharedVault, to, amount);
+
+        emit Withdrawn(recipient, to, amount);
+    }
+
+    function _voidGrant(Grant storage g)
+        private
+        returns (uint128 amountVoided)
+    {
+        uint128 amountOwned = _getAmountOwned(g);
+
+        amountVoided = g.amount - amountOwned;
+        g.amount = amountOwned;
+
+        g.grantStart = 0;
+        g.grantPeriod = 0;
+    }
+
+    function _getAmountOwned(Grant memory g) private view returns (uint128) {
+        return _calcAmount(g.amount, g.grantStart, g.grantCliff, g.grantPeriod);
+    }
+
+    function _getAmountUnlocked(Grant memory g)
+        private
+        view
+        returns (uint128)
+    {
+        return _calcAmount(
+            _getAmountOwned(g), g.unlockStart, g.unlockCliff, g.unlockPeriod
+        );
+    }
+
+    function _calcAmount(
+        uint128 amount,
+        uint64 start,
+        uint64 cliff,
+        uint64 period
+    )
+        private
+        view
+        returns (uint128)
+    {
+        if (amount == 0) return 0;
+        if (start == 0) return amount;
+        if (block.timestamp <= start) return 0;
+
+        if (period == 0) return amount;
+        if (block.timestamp >= start + period) return amount;
+
+        if (block.timestamp <= cliff) return 0;
+
+        return amount * uint64(block.timestamp - start) / period;
```

Comment:
Also a logical issue here: `amount * uint64(block.timestamp - start) / period;` with an exmaple.

Bob owns 100_000 tokens.
Bob waits until it is fully unlocked + some more time.
After it fully unlocked, and he waits more, he could get a multiplier because of this: `* uint64(block.timestamp - start) / period;`
Best to hard-cap that that amount somewhere (either here or at `transferFrom()`)

Reply:
@dantaik `if (block.timestamp >= start + period) return amount;` eliminates this. (No issue here then :) )